taichi/ir/frontend_ir.cpp

```diff
@@ -173,6 +173,29 @@ void UnaryOpExpression::flatten(FlattenContext *ctx) {
   ctx->push_back(std::move(unary));
 }
 
+Expr to_broadcast_tensor(const Expr &elt, const DataType &dt) {
+  TI_ASSERT(dt->is<TensorType>());
+  auto tensor_type = dt->as<TensorType>();
+  auto elt_type = tensor_type->get_element_type();
+  TI_ASSERT_INFO(elt_type->is<PrimitiveType>(),
+                 "Only primitive types are supported in Tensors, got {}",
+                 elt_type->to_string());
+  std::vector<Expr> broadcast_values(tensor_type->get_num_elements(), elt);
+  return Expr::make<MatrixExpression>(broadcast_values,
+                                      tensor_type->get_shape(), elt_type);
+}
+
+std::tuple<Expr, Expr> unify_binop_operands(const Expr &e1, const Expr &e2) {
+  if ((!e1->ret_type->is<TensorType>() && !e2->ret_type->is<TensorType>()) ||
+      (e1->ret_type->is<TensorType>() && e2->ret_type->is<TensorType>())) {
+    return std::tuple(e1, e2);
+  }
+  if (!e1->ret_type->is<TensorType>()) {
```

Comment:
ditto. `!e1->ret_type->is<TensorType>()` does not guarantee it's PrimitiveType. It could be unexpected types suchas PointerType or some other types added in the future. In that case we better throw an eror.